bowling/bowling.rb

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+MAX_PIN = 10
+MAX_FRAME = 10
+BASIC_SIZE_1FRAME = 2
+
+class Frame
+  def initialize(points, previous_frame = nil, pre_previous_frame = nil)
+    @points = points
+    @previous_frame = previous_frame
+    @pre_previous_frame = pre_previous_frame
+  end
+
+  def score
+    @points.sum + score_spare + score_strike
+  end
+
+  def score_spare
+    if !@previous_frame.nil? && @previous_frame.spare?
+      @points[0] # スペアなら1フレームの一投目だけを追加で加算
+    else
+      0
+    end
+  end
+
+  def score_strike
+    if !@previous_frame.nil? && @previous_frame.strike?
+      sum = @points[0] # 前のフレームがストライクならまずは1フレームの一投目だけを追加で加算
+      sum += @points[1] if @points.size >= BASIC_SIZE_1FRAME # ストライクまたは半フレーム（フレームの途中）以外なら2投目を追加で加算
+      sum += @points[0] if !@pre_previous_frame.nil? && @pre_previous_frame.strike? # 前と2個前がストライクなら更に1投目を追加で加算
+      sum
+    else
+      0
+    end
+  end
+
+  def spare?
+    @points.size == BASIC_SIZE_1FRAME && @points.sum == MAX_PIN
+  end
+
+  def strike?
+    @points.size == BASIC_SIZE_1FRAME - 1 && @points[0] == MAX_PIN
+  end
+end
+
+def pin_data2int(input)
+  input.gsub('X', MAX_PIN.to_s).split(',').map(&:to_i) # X->10に置換
+end
+
+def separate_frame(pins)
+  all_frame = [] # 全フレームの倒したピンの数の配列（１フレームごとに入れ子の配列になっている）
+  pins_1frame = [] # 1フレームの倒したピンの数の配列
+
+  pins.each do |pin|
+    pins_1frame << pin
+    maxsize_1frame = all_frame.size >= MAX_FRAME - 1 ? BASIC_SIZE_1FRAME + 1 : BASIC_SIZE_1FRAME # 最後のフレームはBASIC_SIZE_1FRAME + 1投げられる
```

Comment:
[IMO]

フレームの計算ごとに`all_frame`の要素数が増えていき要素数をもとに条件判別していますが、
例えば、2フレーム目の計算では`all_frame.sieze`が1となっており、1ずれてしまっているのが少し分かりづらいと感じました。

そこで、`with_index`メソッドを使うのはいかがでしょうか？
https://docs.ruby-lang.org/ja/latest/method/Enumerator/i/with_index.html

```suggestion
  pins.each.with_index(1) do |pin, index|
    pins_1frame << pin
    maxsize_1frame = index >= MAX_FRAME ? BASIC_SIZE_1FRAME + 1 : BASIC_SIZE_1FRAME # 最後のフレームはBASIC_SIZE_1FRAME + 1投げられる
```
のように計算しているフレーム数を`index`で判別することができます。

Reply:
ごめんなさい、フレームごとではなく`pins`ごとに`index`が増えてしまったので、使えませんでした 🙇

Reply:
参考までに、私だったらどう実装するか考えてみました。

```rb
def seperate_frame(pins)
    def next_frame?(pairs)
        return true if pairs.size == 0
        return false if pairs.size == 10
    
        pairs.last.size == 2 || pairs.last.first == 10
    end

    pins.inject([]) do |pairs, pin|
        frame_count = next_frame?(pairs) ? pairs.size + 1 : pairs.size

        # 初回の投球
        if pairs.size == 0
            pairs.push([pin])
            next pairs
        end

        # 10フレーム目の場合
        if frame_count == 10 
            pair_for_10 = pairs[9] || [] # 10フレーム目のpair
            pair_for_10.push(pin)
            pairs[9] = pair_for_10
            next pairs
        end

        # 直前のフレームがストライク、または2回投げていたら新しいスコアとして追加する
        # 1回目のみのスコアだったら既存のスコアを更新する
        previous_pair = pairs.pop || []
        if previous_pair.size == 2 || previous_pair[0] == 10
            pairs.push(previous_pair, [pin])
            next pairs
        else
            new_pair = [previous_pair[0], pin]
            pairs.push(new_pair)
            next pairs
        end
    end
end
```

検証結果は以下です。
```rb
p seperate_frame([1,2,3]) # => [[1, 2], [3]]
p seperate_frame([10,2,3]) # => [[10], [2, 3]]
p seperate_frame([1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,1,2]) # => [[1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 6], [7, 7], [8, 8], [9, 9], [10, 1, 2]]
```

Reply:
## separate_frameメソッド内の条件分岐に関して
> 参考までに、私だったらどう実装するか考えてみました

一応、もっと短くしてみました。
もっと、簡潔に書ける方法があれば教えて欲しいです。
```ruby
def separate_frame(all_pins)
  all_pins.each_with_object([[]]) do |pin, pairs|
    if pairs.last.empty? || pairs.size >= MAX_FRAME || (pairs.last.size < BASIC_SIZE_1FRAME && pairs.last.last < MAX_PIN)
      pairs.last << pin
    else
      pairs << [pin]
    end
  end
end
```

テスト
```ruby
p separate_frame([1,2,3]) # => [[1, 2], [3]]
p separate_frame([10,2,3]) # => [[10], [2, 3]]
p separate_frame([1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,1,2]) # => [[1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 6], [7, 7], [8, 8], [9, 9], [10, 1, 2]]
```
実行結果
```
$ /usr/bin/env ruby "/Users/atsushi/Documents/Fjord/yuma_review/bowling/test2.rb"
[[1, 2], [3]]
[[10], [2, 3]]
[[1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 6], [7, 7], [8, 8], [9, 9], [10, 1, 2]]
```

良かったら、全体修正したので見てくれると嬉しい！

## separate_frameやstrike?メソッドのテストに関して
他の課題で以下のレビューを受けたんですが、separate_frameやstrike?メソッドってプライベートなメソッドに当たりますか？

> テストの粒度が細かい & 外部に公開されたメインロジックに対するテストになっていない
> 少し話はそれるようですが、プライベートメソッドに対するテストは書かないものである、という話があります: https://t-wada.hatenablog.jp/entry/should-we-test-private-methods
簡単に言うと、
> - 外部に公開している（外部から利用される・呼び出される）処理が期待通りに動くことの確認をテストの主眼としたほうがよく、その処理の実装の詳細に対するテストは書かない
> - 実装の詳細は、外部に公開している処理を通じてテストできるため
>
> ということが言われています。

> たぶん、最終的にプライベートになる部分に対してもテストを先に作成する、というのがちょっと違っていそうです👀 例えば 「TDD の基本を理解する」で引用されている [appkitbox.com/knowledge/test/20130115-367512](https://appkitbox.com/knowledge/test/20130115-367512) を見てみても、記事中で実装していく isLeapYear 関数自体に対するテストを書いています。「プライベートになる部分に対」するテストというのは、 isLeapYear の中で呼び出される year % 4 == 0 や year % 100 != 0 に対するテストということになりますが、そういったテストは書いてありません。
あくまでも、終始 isLeapYear に対するテストを書き、そのテストが内部の実装を通るようにテストケースをふやしていく、というのがテストのやり方です。それは TDD でもそうでなくても変わらない、というのが僕の考えです！